File: mobile/lib/domain/usecases/submit_placement_test_use_case.dart
import '../repositories/i_course_repository.dart';
import '../repositories/i_learner_repository.dart';
import '../entities/question.dart';
import '../entities/skill_score.dart';
import '../personalization/skill_scoring_engine.dart';

class AssessmentResult {
  final String assessmentId;
  final String courseId;
  final List<QuestionAnswerSubmission> submissions;
  final Map<String, SkillScore> computedScores;
  final String weakestSkillId;
  final int timestamp;

  const AssessmentResult({
    required this.assessmentId,
    required this.courseId,
    required this.submissions,
    required this.computedScores,
    required this.weakestSkillId,
    required this.timestamp,
  });
}

class SubmitPlacementTestUseCase {
  final ICourseRepository courseRepository;
  final ILearnerRepository learnerRepository;

  const SubmitPlacementTestUseCase(this.courseRepository, this.learnerRepository);

  Future<AssessmentResult> execute(String courseId, List<QuestionAnswerSubmission> submissions) async {
    final questions = await courseRepository.getPlacementQuestions(courseId);

    // Record answer attempts in repository
    for (final sub in submissions) {
      final q = questions.where((item) => item.id == sub.questionId).firstOrNull;
      if (q != null) {
        final optionText = sub.selectedIndex < q.options.length ? q.options[sub.selectedIndex] : '';
        await learnerRepository.recordAnswerAttempt(
          courseId: courseId,
          skillId: q.skillId,
          lessonId: 'placement_test',
          questionId: q.id,
          selectedIndex: sub.selectedIndex,
          selectedAnswer: optionText,
          isCorrect: sub.isCorrect,
        );
      }
    }

    // Real statistical placement scoring
    final computedScores = SkillScoringEngine.computePlacementScores(courseId, questions, submissions);

    // Save to learner profile
    await learnerRepository.saveSkillScores(courseId, computedScores);

    // Award +20 XP for completing placement test (idempotent)
    await learnerRepository.addXp(20, 'placement_completed_$courseId');
    await learnerRepository.recordDailyActivity();

    // Identify weakest skill dynamically
    String weakestSkillId = '';
    int minScore = 101;
    for (final s in computedScores.values) {
      if (s.score < minScore) {
        minScore = s.score;
        weakestSkillId = s.skillId;
      }
    }

    return AssessmentResult(
      assessmentId: 'placement_$courseId',
      courseId: courseId,
      submissions: submissions,
      computedScores: computedScores,
      weakestSkillId: weakestSkillId,
      timestamp: DateTime.now().millisecondsSinceEpoch,
    );
  }
}

File: src/domain/usecases/SubmitPlacementTestUseCase.ts
import { ICourseRepository } from '../repositories/ICourseRepository';
import { ILearnerRepository } from '../repositories/ILearnerRepository';
import { QuestionAnswerSubmission } from '../entities/Question';
import { AssessmentResult } from '../entities/Assessment';
import { SkillScoringEngine } from '../personalization/SkillScoringEngine';

export class SubmitPlacementTestUseCase {
  constructor(
    private courseRepo: ICourseRepository,
    private learnerRepo: ILearnerRepository
  ) {}

  async execute(courseId: string, submissions: QuestionAnswerSubmission[]): Promise<AssessmentResult> {
    const questions = await this.courseRepo.getPlacementQuestions(courseId);

    // Record answer attempts for audit and logging
    for (const sub of submissions) {
      const q = questions.find((item) => item.id === sub.questionId);
      if (q) {
        await this.learnerRepo.recordAnswerAttempt({
          courseId,
          skillId: q.skillId,
          lessonId: 'placement_test',
          questionId: q.id,
          selectedIndex: sub.selectedIndex,
          selectedAnswer: q.options[sub.selectedIndex] || '',
          isCorrect: sub.isCorrect,
        });
      }
    }

    // Compute real scores strictly from actual submissions
    const computedScores = SkillScoringEngine.computePlacementScores(courseId, questions, submissions);
    const overallScore = SkillScoringEngine.computeOverallScore(computedScores);

    // Find weakest skill dynamically
    let weakestSkillId = '';
    let minScore = 101;
    Object.values(computedScores).forEach((s) => {
      if (s.score < minScore) {
        minScore = s.score;
        weakestSkillId = s.skillId;
      }
    });

    // Save placement attempt & individual answers to Supabase / repository
    const attemptId = await this.learnerRepo.savePlacementAttempt({
      courseId,
      score: overallScore,
      weakestSkillId,
      submissions: submissions.map((s) => ({
        questionId: s.questionId,
        selectedIndex: s.selectedIndex,
        isCorrect: s.isCorrect,
      })),
    });

    // Save scores into learner repository
    await this.learnerRepo.saveSkillScores(courseId, computedScores);

    // Award +20 XP for completing placement test (idempotent)
    await this.learnerRepo.addXp(20, `placement_completed_${courseId}`);
    await this.learnerRepo.recordDailyActivity();

    return {
      assessmentId: attemptId || `placement_${courseId}`,
      courseId,
      submissions,
      computedScores,
      weakestSkillId,
      timestamp: Date.now(),
    };
  }
}
